shared/runtime/pyexec.c

```diff
@@ -175,6 +189,11 @@ static int parse_compile_execute(const void *source, mp_parse_input_kind_t input
         mp_hal_stdout_tx_strn("\x04", 1);
     }
 
+    #if MICROPY_PY_ATEXIT
+    int atexit_code = mp_atexit_execute();
```

Comment:
I haven't tested this, but placing this here in `parse_compile_execute()` seems like it might have some unexpected consequences? For example, if you put an `atexit.register()` call in `boot.py` then will the handler be run after executing `boot.py` and before executing `main.py`?

(I think this draws attention to a shortcoming in the current architecture, which is that AFAIK there's no non-hardware-specific soft reset hook function that all ports can call for things which need cleanup in the soft reset path.)

Reply:
Yes I've realised similar issues with this location and been testing / looking at other ways to add this cleanly.

I've actually been fighting the urge to undertake a broad refactor & consolidation of soft-reset / shutdown handling in ports to create the "correct" place to run the handler code! I'd rather not block this change behind such a large refactor though. 

Perhaps for now I should go back to my earlier plan to just add the execute call to an appropriate place in each main.c 
Then they'll all get cleaned up anyway if I get to the refactor.

Reply:
> Perhaps for now I should go back to my earlier plan to just add the execute call to an appropriate place in each main.c Then they'll all get cleaned up anyway if I get to the refactor.

That sounds like a good approach to me. FWIW, I have also felt this urge... but better to do it as a pure refactor, for sure.